src/frequenz/core/enum.py

```diff
@@ -0,0 +1,204 @@
+# License: MIT
+# Copyright © 2025 Frequenz Energy-as-a-Service GmbH
+
+"""Enum utilities with support for deprecated members.
+
+This module provides an [`Enum`][frequenz.core.enum.Enum] base class that extends the
+standard library's [`enum.Enum`][] to support marking certain members as deprecated.
+
+See the [class documentation][frequenz.core.enum.Enum] for details and examples.
+"""
+
+from __future__ import annotations
+
+import enum
+import warnings
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Any, ClassVar, Self, TypeVar, cast
+
+# Note: This module contains more casts and uses of Any than what's typically
+# ideal. This is because type hinting EnumMeta and Enum subclasses is quite
+# challenging, as there is a lot of special behavior in `mypy` for these classes.
+#
+# The resulting enum should be treated as a regular enum by mypy, so hopefully everthing
+# still works as expected.
+
+EnumT = TypeVar("EnumT", bound=enum.Enum)
+"""Type variable for enum types."""
+
+
+class DeprecatedMemberWarning(DeprecationWarning):
+    """Warning category for deprecated enum members."""
+
+
+class DeprecatedMember:
+    """Marker used in enum class bodies to declare deprecated members.
+
+    Please read the [`Enum`][frequenz.core.enum.Enum] documentation for details and
+    examples.
+    """
+
+    __slots__ = ("value", "message")
+
+    def __init__(self, value: Any, message: str) -> None:
+        """Initialize this instance."""
+        self.value = value
+        self.message = message
+
+
+class DeprecatingEnumMeta(enum.EnumMeta):
+    """Enum metaclass that supports `DeprecatedMember` wrappers.
+
+    Tip:
+        Normally it is not necessary to use this class directly, use
+        [`Enum`][frequenz.core.enum.Enum] instead.
+
+    Behavior:
+
+    - In the class body, members may be declared as `NAME = DeprecatedMember(value, msg)`.
+    - During class creation, these wrappers are replaced with ``value`` so that
```

Comment:
Is the double ` deliberate here?

Reply:
No, good catch, it was AI thinking we use reST, will fix.

Reply:
Fixed in fixup commit